src/task.rs

```diff
@@ -396,37 +351,76 @@ impl Task {
             return self.notify_starling_errored(Error::from_kind(ErrorKind::JavaScriptException));
         }
 
-        // Drain the micro-task queue.
-
         unsafe {
             jsapi::js::RunJobs(cx);
+        }
 
-            if jsapi::JS_IsExceptionPending(cx) {
+        if let Err(e) = self.check_for_unhandled_rejected_promises() {
+            return self.notify_starling_errored(e);
+        }
+
+        self.evaluate_main()
+    }
+
+    fn evaluate_main(&mut self) -> Result<Async<()>> {
+        let cx = self.runtime().cx();
+        rooted!(in(cx) let global = self.global.get());
+
+        let mut has_main = false;
+        unsafe {
+            assert!(jsapi::JS_HasProperty(
+                cx,
+                global.handle(),
+                b"main\0".as_ptr() as _,
+                &mut has_main
+            ));
+        }
+        if !has_main {
+            return self.notify_starling_finished();
+        }
+
+        rooted!(in(cx) let mut rval = jsval::UndefinedValue());
+        let args = jsapi::JS::HandleValueArray::new();
+        unsafe {
+            let ok = jsapi::JS_CallFunctionName(
+                cx,
+                global.handle(),
+                b"main\0".as_ptr() as _,
+                &args,
+                rval.handle_mut()
+            );
+
+            if !ok {
+                debug_assert!(jsapi::JS_IsExceptionPending(cx));
                 // TODO: convert the pending exception into a meaningful error.
                 jsapi::JS_ClearPendingException(cx);
+                jsapi::js::RunJobs(cx);
```

Comment:
hmm, I'm not sure it's obvious we should run jobs here. It's also not obvious we shouldn't, so perhaps just add a TODO saying that we need to figure that out?